// @mswjs/interceptors/lib/interceptors/XMLHttpRequest/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.XMLHttpRequestInterceptor = void 0;
var Interceptor_1 = require("../../Interceptor");
var XMLHttpRequestOverride_1 = require("./XMLHttpRequestOverride");
var XMLHttpRequestInterceptor = /** @class */ (function (_super) {
    __extends(XMLHttpRequestInterceptor, _super);
    function XMLHttpRequestInterceptor() {
        return _super.call(this, XMLHttpRequestInterceptor.symbol) || this;
    }
    XMLHttpRequestInterceptor.prototype.checkEnvironment = function () {
        return (typeof window !== 'undefined' &&
            typeof window.XMLHttpRequest !== 'undefined');
    };
    XMLHttpRequestInterceptor.prototype.setup = function () {
        var log = this.log.extend('setup');
        log('patching "XMLHttpRequest" module...');
        var PureXMLHttpRequest = window.XMLHttpRequest;
        window.XMLHttpRequest = XMLHttpRequestOverride_1.createXMLHttpRequestOverride({
            XMLHttpRequest: PureXMLHttpRequest,
            emitter: this.emitter,
            log: this.log,
        });
        log('native "XMLHttpRequest" module patched!', window.XMLHttpRequest.name);
        this.subscriptions.push(function () {
            window.XMLHttpRequest = PureXMLHttpRequest;
            log('native "XMLHttpRequest" module restored!', window.XMLHttpRequest.name);
        });
    };
    XMLHttpRequestInterceptor.symbol = Symbol('xhr');
    return XMLHttpRequestInterceptor;
}(Interceptor_1.Interceptor));
exports.XMLHttpRequestInterceptor = XMLHttpRequestInterceptor;
